### ISSUE TITLE
Handle database corruption while using the database.

### ISSUE THREAD

**Issue by mhammond** [mozilla/application-services#4639]
#2628 is about databases which are unable to be opened due to corruption. This issue is about databases which are able to be opened, but report corruption during normal use of the database. These databases are likely to be able to be recovered.

@mak77 told me in slack:
> When you can open the database and PRAGMA integrity_check() returns some errors, one can try to REINDEX, and I noticed 90% of the times that is sufficient.
> Of course it won't fix actual file corruption. For that what we try to do, atm, is copying entities one by one to a new database.
> If you can't open the db at all, there's not much you can do for now, there is a corruption in the main header of the database.
> I must note that recently Sqlite has added a Recovery API to try recovering data from broken databases, see https://www.sqlite.org/recovery.html. We didn't implement it yet, and we are blocked on Sqlite updates because of [bug 1781116](https://bugzilla.mozilla.org/show_bug.cgi?id=1781116), we tried to ping multiple times to have a fix for that, but it's not being fixed. So if you want to try using that API, you'll likely have to escalate that issue.

Some other notes about how desktop currently does it (which might not be completely relevant given the above, but I guess might be if the above process fails)

* Desktop makes an [attempt to repair it](https://searchfox.org/mozilla-central/rev/81fbf5597044b6811e6aa4a10f3d78e8c3080f8c/toolkit/components/places/PlacesDBUtils.sys.mjs#110-128) by [doing the REINDEX described above](https://searchfox.org/mozilla-central/rev/81fbf5597044b6811e6aa4a10f3d78e8c3080f8c/toolkit/components/places/PlacesDBUtils.sys.mjs#1443).
*  If that fails, it does a more complex recovery step as implemented in [bug 1411538](https://bugzilla.mozilla.org/show_bug.cgi?id=1411538), which the tl;dr is:
* A preference is set and the database remains broken until the next restart.
* On the next startup when that preference is set, is tries to manually recover as much data as it can, as implemented [here](https://searchfox.org/mozilla-central/source/toolkit/components/places/Database.cpp#735) 

If we need to do things like desktop does, it's not clear we will be able to implement this only once as the schema for each of our databases is different. We should be able to reuse much code between them though.

Note the original description of this issue said "We now handle corrupt databases while opening the DB and have a utility function most components use to handle this" - but we do not!

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2884)

**Comment by mhammond** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1865193

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).
